lib/api/src/lib/stories.ts

```diff
@@ -144,103 +136,89 @@ export const transformStoriesRawToStoriesHash = (
   input: StoriesRaw,
   { provider }: { provider: Provider }
 ): StoriesHash => {
-  const anyKindMatchesOldHierarchySeparators = Object.values(input)
-    .filter(Boolean)
-    .some(({ kind }) => kind.match(/\.|\|/));
+  const values = Object.values(input).filter(Boolean);
+  const usesOldHierarchySeparator = values.some(({ kind }) => kind.match(/\.|\|/)); // dot or pipe
 
-  const storiesHashOutOfOrder = Object.values(input)
-    .filter(Boolean)
-    .reduce((acc, item) => {
-      const { kind, parameters } = item;
-      const { showRoots } = provider.getConfig();
+  const storiesHashOutOfOrder = values.reduce((acc, item) => {
+    const { kind, parameters } = item;
+    const { showRoots } = provider.getConfig();
 
-      const setShowRoots = typeof showRoots !== 'undefined';
-      if (anyKindMatchesOldHierarchySeparators && !setShowRoots) {
-        warnChangedDefaultHierarchySeparators();
-      }
+    const setShowRoots = typeof showRoots !== 'undefined';
+    if (usesOldHierarchySeparator && !setShowRoots) {
+      warnChangedDefaultHierarchySeparators();
+    }
 
-      let root = '';
-      let groups: string[];
-      const parts: string[] = kind.split('/');
-      // Default showRoots to true if they didn't set it.
-      if ((!setShowRoots || showRoots) && parts.length > 1) {
-        [root, ...groups] = parts;
-      } else {
-        groups = parts;
-      }
+    const groups = kind.split('/').map((part) => part.trim());
+    const root = (!setShowRoots || showRoots) && groups.length > 1 ? [groups.shift()] : [];
+
+    const rootAndGroups = [...root, ...groups].reduce((list, name, index) => {
+      const parent = index > 0 && list[index - 1].id;
+      const id = sanitize(parent ? `${parent}-${name}` : name);
 
-      const rootAndGroups = []
-        .concat(root || [])
-        .concat(groups)
-        .map(toGroup)
-        // Map a bunch of extra fields onto the groups, collecting the path as we go (thus the reduce)
-        .reduce((soFar, group, index, original) => {
-          const { name } = group;
-          const parent = index > 0 && soFar[index - 1].id;
-          const id = sanitize(parent ? `${parent}-${name}` : name);
-          if (parent === id) {
-            throw new Error(
-              dedent`
+      if (parent === id) {
+        throw new Error(
+          dedent`
               Invalid part '${name}', leading to id === parentId ('${id}'), inside kind '${kind}'
 
               Did you create a path that uses the separator char accidentally, such as 'Vue <docs/>' where '/' is a separator char? See https://github.com/storybookjs/storybook/issues/6128
             `
-            );
-          }
+        );
+      }
 
-          if (!!root && index === 0) {
-            const result: Root = {
-              ...group,
```

Comment:
Note `group` had only two properties, `name` and `id`. But `id` is overwritten immediately after, so effectively we're only assigning `name` here. Might as well do that directly.